### PR TITLE
I/3287: Table and table cell properties plugins

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduce table properites conversion. Closes ckeditor/ckeditor5#6048.

---
## PR constallation

Build on ckeditor5/i3287 branch: [Build Status](https://travis-ci.org/ckeditor/ckeditor5)

_Edit_: 
~the build fails because watchdog detects `StylesConverter` as "connected" https://github.com/ckeditor/ckeditor5-watchdog/blob/73844f37f90ea0bf0e0a756752c62b1ce86e334b/src/watchdog.js#L452 because of the `StylesConverter` singleton...~ it is fixed by making StyleConverter a non-enumerable property.

This PR requires changes:

- `engine`: https://github.com/ckeditor/ckeditor5-engine/pull/1807 (comes with own constellation of changes)
- ~`ui`: https://github.com/ckeditor/ckeditor5-ui/pull/528 (might be reverted as UI is not a part of MVP)~

## Description

This is a draft PR to start review process of this behemoth change. This PR will need at least two reviewers and probably more then 2 passes.

At this stage I'd like to review the general idea behind table styles:

- Changes in the engine that enables styles normalization (separate PR)
- Table properties model, conversion, desired output.

Conversion (`*` - means any value but can be constrained by `Styles` normalizer):

1. `TableProperties` handles:
    - `borderWidth=*`, `borderColor=*`, `borderStyle=*` (`<table style="boder-top:1px solid blue;...">`):
    - `backgroundColor=*` (`<table style="background-color:some-color">`)
    - `width=*` / `heigh=*` ( `<table style="width:xx">`)
    - `alignment=left|right|center` (`<table style="margin-left:0,margin-right:auto">` - as in [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table#attr-align), upcasts also `<table aling="left|right|center">`).
2. `TableCellProperties` handles:
    - `borderWidth=*`, `borderColor=*`, `borderStyle=*` (`<td style="boder-top:1px solid blue;...">`)
    - `backgroundColor=*` (`<td style="background-color:some-color">`)
    - `padding=*` (`<td style="padding:2em">`)
    - `verticalAlignment=*` (`<td style="vertical-align:top">`)  - should be fixed to
    - `horizontalAlignment=left|fight|center|justify` (`<td style="text-align:left|right|center"justify">`)
3. `TableColumnRowProperties` handles:
    - `heigth=*` (`<td style="height:*">`)
    - `width=*` (`<td style="widith:*">`)

### CODE REVIEW

**Comment by ConnorSinnott** [comment]
Hey @jodator, @Reinmar. Its a bummer to see this has been bumped off the December milestone. In the meantime I think I am going to have to piece together a temporary solution for table styling as we are looking to implement CKEditor5 as a CKEditor4 replacement come January. As far as I can tell there have not been any UI implemented for this feature yet. Did you have any thoughts on what the UI would look like? I want to make sure the transition from our implementation to yours is as seamless as possible. 

image

Thank you!

**Comment by jodator** [comment]
Hi @ConnorSinnott, unfortunately, we need to push with other features and the changes behind this requires careful handling in the editor engine itself. AFAIR we don't have any UI ticket for this but when we start working on it I'll link to your proposal and cc you.

**Comment by ConnorSinnott** [comment]
@jodator I completely understand, and please do keep me in the loop, I would definitely appreciate it. I'll also be sure to submit whatever we land on here to your team as a proposal. Thank you

**Comment by ConnorSinnott** [comment]
Hey all! This is the direction I think we are going to end up taking things. We realized that without multi-cell select, just having the ability to color a single cell would not be very useful. In the meantime, we decided to piggyback on the 'Row' and 'Column' management tools, and added the ability to color an entire row, or an entire column. 

Here is what we are working with atm.
ColorTable

Peek 2019-12-12 14-47

**Comment by remigasas** [comment]
Hello,
we are also starting to use this plugin changes without them being officially release.
After merging these changes and building ck I'm getting this error:

```
TypeError: data.viewItem.getNormalizedStyle is not a function
    at UpcastDispatcher.<anonymous> (ckeditor.js:70113)
    at UpcastDispatcher.fire (ckeditor.js:89305)
    at UpcastDispatcher._convertItem (ckeditor.js:12960)
    at UpcastDispatcher._convertChildren (ckeditor.js:12991)
    at UpcastDispatcher.<anonymous> (ckeditor.js:13983)
    at UpcastDispatcher.fire (ckeditor.js:89305)
    at UpcastDispatcher._convertItem (ckeditor.js:12960)
    at ckeditor.js:69358
    at Array.forEach (<anonymous>)
    at UpcastDispatcher.<anonymous> (ckeditor.js:69358)
```
It coming from ckeditor5-table\src\tableproperties\utils.js line 53.
Yes, this is my mistake for not taking all changes in other PR (I think it is related to PR in other projects).

This story is not yet release, and it has been delayed. I would like to ask if there a plan to merge changes bit by bit ( at least those global in other projects ) ? I'm looking at several of those in-progress PR's and it becomes difficult to track and build ckeditor from them.

**Comment by ConnorSinnott** [comment]
@remigasas the `getNormalizedStyle` is a method not yet induced to the CKeditor5 engine, but the pull request can be found [here](https://github.com/ckeditor/ckeditor5-engine/pull/1807). In it's absence, you might be able to change the method to `getStyle` and see if that will works for you. Just know that at the moment this branch is not usable without heavy modifications.

**Comment by jodator** [comment]
@Reinmar This PR implements the two main plugins:

1. `TableCellProperties` - glues `*Editing` and `*UI` parts.
1. `TableProperties` - glues `*Editing` and `*UI` parts.

From that I've implemented 16 commands:

For table properties:

- `tableBorderColor` - manipulates a `table`'s `borderColor` attribute
- `tableBorderStyle` - manipulates a `table`'s `borderStyle` attribute
- `tableBorderWidth` - manipulates a `table`'s `borderWidth` attribute
- `tableAlignment` - manipulates a `table`'s `alignment` attribute
- `tableWidth` - manipulates a `table`'s `width` attribute
- `tableHeight` - manipulates a `table`'s `height` attribute
- `tableBackgroundColor` - manipulates a `table`'s `backgroundColor` attribute

For table cell properties:

- `tableCellBorderStyle` - manipulates a `tableCell`'s `borderStyle` attribute
- `tableCellBorderColor` - manipulates a `tableCell`'s `borderColor` attribute
- `tableCellBorderWidth` - manipulates a `tableCell`'s `borderWidth` attribute
- `tableCellHorizontalAlignment` - manipulates a `tableCell`'s `horizontalAlignment` attribute
- `tableCellWidth` - manipulates a `tableCell`'s `width` attribute
- `tableCellHeight` - manipulates a `tableCell`'s `height` attribute
- `tableCellPadding` - manipulates a `tableCell`'s `padding` attribute
- `tableCellBackgroundColor` - manipulates a `tableCell`'s `backgroundColor` attribute
- `tableCellVerticalAlignment` - manipulates a `tableCell`'s `verticalAlignment` attribute

Please validate if the names are OK (especially if the attribute name is OK ie `borderColor` vs `tableBorderColor`). Eventual change would be 20 min tops, so thats not an issue ;)

Most of the commands code is done in base commands which at least in MVP do the same: get a table or cells from selection and apply or remove attribute. If command operates on top/right/bottom/left value object it will set `command.value` only if all four are the same.

As a simplification (which also works) I've made that command sets one string value in the model instead of `top/right/bottom/left` object. This can be changed in follow up.

File: tests/manual/tableproperties.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+### Loading
```
[the remark this answers is not in the document]
Comment (jodator) [reply tests/manual/tableproperties.md]: Could you check the manual sample again? I've revamped it a bit to make comparing original data to editor rendering easier (side-by-side).

**Comment by jodator** [comment]
@oleq done. Please note that I've added small addition to the docs in the engine so this PR is needed also: https://github.com/ckeditor/ckeditor5-engine/pull/1816.

File: src/commands/utils.js [hunk 2]
```diff
@@ -58,6 +58,22 @@ export function createEmptyTableCell( writer, insertPosition, attributes = {} )
 	writer.insert( tableCell, insertPosition );
 }
 
+/**
+ * Returns a string if all four values of box edges are equal.
```
Comment (oleq) [inline src/commands/utils.js]:
```suggestion
 * Returns a string if all four values of box sides are equal.
```

File: src/commands/utils.js [hunk 3]
```diff
@@ -58,6 +58,22 @@ export function createEmptyTableCell( writer, insertPosition, attributes = {} )
...
+ *
+ * If string is passed it is treated as a single value (pass-through).
```
Comment (oleq) [inline src/commands/utils.js]:
```suggestion
 * If a string is passed, it is treated as a single value (pass-through).
```

**Pull request merged**